sycl/include/CL/sycl/ONEAPI/sub_group.hpp

```diff
@@ -699,13 +700,40 @@ struct sub_group {
 #endif
   }
 
-protected:
-  template <int dimensions> friend class cl::sycl::nd_item;
-  friend sub_group this_sub_group();
-  sub_group() = default;
-};
+  linear_id_type get_group_linear_range() const {
+#ifdef __SYCL_DEVICE_ONLY__
+    return static_cast<linear_id_type>(get_group_range()[0]);
+#else
+    throw runtime_error("Sub-groups are not supported on host device.",
+                        PI_INVALID_DEVICE);
+#endif
+  }
+
+  linear_id_type get_local_linear_range() const {
+#ifdef __SYCL_DEVICE_ONLY__
+    return static_cast<linear_id_type>(get_local_range()[0]);
+#else
+    throw runtime_error("Sub-groups are not supported on host device.",
+                        PI_INVALID_DEVICE);
+#endif
+  }
+
+  bool leader() const {
+#ifdef __SYCL_DEVICE_ONLY__
+    return detail::spirv::GroupNonUniformElect<sub_group>();
```

Comment:
```suggestion
    return 0;
```

Your implementation is more complicated than it needs to be, and may lead developers to the incorrect conclusion that they can rely on `GroupNonUniformElect` behavior inside of control flow.  The SYCL 2020 specification says:

> The leader of the sub-group is determined during construction of the sub-group, and is invariant for the lifetime of the sub-group. The leader of the sub-group is guaranteed to be the work-item with a local id of 0.

Reply:
@Pennycook should it be 
```
  return get_local_linear_id() == 0;
```
then? Otherwise we have no leader work-item.

Reply:
Sorry, yes -- I forgot that this returns a `bool`!  Your code is right, mine is wrong.